lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,182 @@
+require 'nokogiri'
+require 'htmlentities'
+
+class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
+  IGNORABLE_TAGS = %w[html head body].freeze
+
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  # http://www.456bereastreet.com/archive/201005/void_empty_elements_and_self-closing_start_tags_in_html/
+  VOID_TAGS = %w[area base br col command hr img input keygen link meta param source].freeze
+
+  attr_reader :truncated_string,
+    :max_length,
+    :tail,
+    :ignored_levels,
+    :something_has_been_truncated
+
+  def initialize(options)
+    @html_coder = HTMLEntities.new
+
+    @max_length = options[:max_length]
+    @tail = options[:tail]
+
+    @truncated_string = ""
+    @closing_tags = []
+    @estimated_length = 0
+    @ignored_levels = 0
+  end
+
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    return if max_length_reached || ignorable_tag?(name)
+
+    string_to_add = opening_tag(name, attributes)
+
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      enter_ignored_level
+      return
+    end
+
+    # Save the tag so we can push it on at the end
+    @closing_tags.push name unless single_tag_element? name
+
+    append_to_truncated_string(string_to_add, length_of_tags)
+  end
+
+  # This method is called when the parser encounters characters between tags
+  def characters(decoded_string)
+    if max_length_reached || ignore_mode?
+      @something_has_been_truncated = max_length_reached
```

Comment:
I'm not sure I understand the need for `@something_has_been_truncated` either.. it seems to only be used in a few places within this one function..the only place we actually check its value is [here](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R61) but we set it to `true` 2 lines before so it's always true by the time we get there..